### PR TITLE
Improve performance in `ActiveStorage::Service::MirrorService`

### PR DESCRIPTION
**Pull request by heka1024**
### Motivation / Background

This Pull Request has been created to remove `FIXME` comment of `ActiveStorage::Service::MirrorService`.

### Detail

This Pull Request use thread pool to parallelize `delete` and `delete_prefixed` in `ActiveStorage::Service::MirrorService`.

### Additional information

If this is a nit performance boost than complexity, how about remove comment?

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activestorage/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+*   Improve performance by parallelizing `ActiveStorage::Service::MirrorService#delete` and `ActiveStorage::Service::MirrorService#delete_prefixed`.
+
+    *heka1024*
```
Comment (akhilgkrishnan) [inline activestorage/CHANGELOG.md:3]: We don't really need to add Changelogs for refactorings
ref: https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#updating-the-changelog

File: activestorage/lib/active_storage/service/mirror_service.rb [hunk 2]
```diff
@@ -30,6 +30,13 @@ def self.build(primary:, mirrors:, name:, configurator:, **options) # :nodoc:
 
     def initialize(primary:, mirrors:)
       @primary, @mirrors = primary, mirrors
+      @executor = Concurrent::ThreadPoolExecutor.new(
```
Comment (akhilgkrishnan) [inline activestorage/lib/active_storage/service/mirror_service.rb:33]: Can you do a performance benchmarking?

File: activestorage/CHANGELOG.md (hunk 1 above)
Comment (heka1024) [reply activestorage/CHANGELOG.md:3]: Thank you for notice!

File: activestorage/lib/active_storage/service/mirror_service.rb (hunk 2 above)
Comment (heka1024) [reply activestorage/lib/active_storage/service/mirror_service.rb:33]: I'll add performance benchmark and request review to you.

File: activestorage/CHANGELOG.md (hunk 1 above)
Comment (heka1024) [reply activestorage/CHANGELOG.md]: I just removed changelog for this PR.

File: activestorage/lib/active_storage/service/mirror_service.rb (hunk 2 above)
Comment (heka1024) [reply activestorage/lib/active_storage/service/mirror_service.rb:33]: @akhilgkrishnan 
This is performance benchmark using GCS. In my opinion, there is a performance improvement when looking at real time.

- repeat time: 10
- Used service: Google Cloud Storage (region: ap-northeast-3)
- Used image size: 512kb

- primary: 1, mirrors: 2

| Task       | Real Time | User Time | System Time | Total Time |
|------------|-----------|-----------|-------------|------------|
| Parallel   | 0.453     | 0.055     | 0.015       | 0.070      |
| Sequential | 0.979     | 0.045     | 0.011       | 0.056      |

- primary: 1, mirrors: 1

| Task       | Real Time | User Time | System Time | Total Time |
|------------|-----------|-----------|-------------|------------|
| Parallel   | 0.380     | 0.053     | 0.020       | 0.074      |
| Sequential | 0.624     | 0.033     | 0.008       | 0.041      |

Comment (heka1024) [reply activestorage/lib/active_storage/service/mirror_service.rb:33]: @akhilgkrishnan Gentle ping 😄 Could you please review when you get a chance?

Comment (heka1024) [reply activestorage/lib/active_storage/service/mirror_service.rb:33]: @akhilgkrishnan Gentle ping 🙏

Comment (akhilgkrishnan) [reply activestorage/lib/active_storage/service/mirror_service.rb:33]: Sorry @heka1024 for the late response. This looks good to me. Lets wait someone from core to review.

File: activestorage/lib/active_storage/service/mirror_service.rb [hunk 3]
```diff
@@ -75,10 +82,12 @@ def each_service(&block)
       end
 
       def perform_across_services(method, *args)
-        # FIXME: Convert to be threaded
-        each_service.collect do |service|
-          service.public_send method, *args
+        tasks = each_service.collect do |service|
+          Concurrent::Promise.execute(executor: @executor) do
+            service.public_send method, *args
```
[the remark this answers is not in the document]
Comment (heka1024) [reply activestorage/lib/active_storage/service/mirror_service.rb:87]: @zzak Actually, I found this TODO while reading the code and worked on it. I haven't experienced performance issues in my application.

I think there are two options:
1. Make this configurable and keep the default as the non-threaded version.
2. Close this PR and remove the TODO comment.

What do you think?

[the remark this answers is not in the document]
Comment (zzak) [reply activestorage/lib/active_storage/service/mirror_service.rb:87]: I like to think of opening PRs sometimes as asking a question. Let's consider this PR the form of "should mirror service use threads?" and see what feedback we get.

I would wait for any more feedback before changing it like option 1, I was just thinking of possible scenarios.

If you don't hear back after a while feel free to [reach out](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#get-some-feedback).

Thanks for your PR!

[the remark this answers is not in the document]
Comment (rafaelfranca) [reply activestorage/lib/active_storage/service/mirror_service.rb:87]: I think we should just assume all services are thread-safe. Rails is entirely thread-safe, and we should push the community to be.

**Pull request merged**